src/Process/PointCloudProcessing.js

```diff
@@ -0,0 +1,300 @@
+
+                        if (boundingBoxEnabled) {
+                            if (!elt.obj.boxHelper) {
+                                initBoundingBox(elt, layer);
+                            }
+                            elt.obj.boxHelper.visible = true;
+                            elt.obj.boxHelper.material.color.r = 1 - shouldBeLoaded;
+                            elt.obj.boxHelper.material.color.g = shouldBeLoaded;
+                        }
+                    }
+
+                    elt.obj.geometry.setDrawRange(
+                        0, Math.floor(shouldBeLoaded * elt.obj.geometry.attributes.position.count));
+                    layer.counters.pointCount += elt.obj.realPointCount;
+                    layer.counters.displayedCount += Math.floor(shouldBeLoaded * elt.obj.geometry.attributes.position.count);
+                    elt.obj.material.uniforms.size.value = layer.pointSize;
+                } else if (!elt.promise) {
+                    // TODO:
+                    // - add command cancelation support
+                    // - rework priority
+                    elt.promise = context.scheduler.execute({
+                        layer,
+                        requester: elt,
+                        view: context.view,
+                        priority: 1.0 / elt.name.length, // surfaceOnScreen,
+                        redraw: true,
+                        isLeaf: elt.childrenBitField == 0,
+                    }).then((pts) => {
+                        elt.obj = pts;
+                        // store tightbbox to avoid ping-pong (bbox = larger => visible, tight => invisible)
+                        elt.tightbbox = pts.tightbbox;
+                        pts.geometry.setDrawRange(
+                            0, shouldBeLoaded * pts.geometry.attributes.position.count);
+
+                        // make sure to add it here, otherwise it might never
+                        // be added nor cleaned
+                        layer.group.add(elt.obj);
+
+                        elt.obj.owner = elt;
+                        elt.promise = null;
+                    });
+                }
+            }
+        } else {
+            // not visible / displayed
+            markForDeletion(elt);
+        }
+
+        if (shouldBeLoaded >= 0.9 && elt.children && elt.children.length) {
+            return elt.children;
+        }
+        return undefined;
+    },
+
+    postUpdate(context, layer) {
+        if (!layer.group) {
+            return;
+        }
+
+        if (layer.counters.displayedCount > layer.pointBudget) {
+            const reduction = layer.pointBudget / layer.counters.displayedCount;
+            for (const pts of layer.group.children) {
+                if (pts.material.visible) {
+                    pts.geometry.setDrawRange(0, pts.geometry.drawRange.count * reduction);
+                }
+            }
+            layer.counters.displayedCount *= reduction;
```

Comment:
Also, hiding the same ratio of points for each node is suboptimal since the points in the nodes higher in the hierarchy are supposed to hold more information. I don't know if this has a big visual impact.